contributors/design-proposals/cpu-manager.md

```diff
@@ -0,0 +1,348 @@
+the other policies are better understood.
+
+#### Policy 2: "static" cpuset control
+
+The "static" policy allocates exclusive CPUs for containers if they are
+included in a pod of "Guaranteed" [QoS class][qos] and the container's
+resource limit for the CPU resource is an integer greater than or
+equal to one.
+
+When exclusive CPUs are allocated for a container, those CPUs are
+removed from the allowed CPUs of every other container running on the
+node. Once allocated at pod admission time, an exclusive CPU remains
+assigned to a single container for the lifetime of the pod (until it
+becomes terminal.)
+
+##### Implementation sketch
+
+```go
+// Implements CPUManagerPolicy
+type staticManager struct {
+  driver CPUDriver
+  topo   CPUTopo
+  // CPU list assigned to non-exclusive containers.
+  shared CPUList
+}
+
+func (m *staticManager) Init(driver CPUDriver, topo CPUTopo) {
+  m.driver = driver
+  m.topo = topo
+}
+
+func (m *staticManager) Add(c v1.Container, qos QoS) error {
+  if p.QoS == GUARANTEED && numExclusive(c) > 0 {
+    excl, err := allocate(numExclusive(c))
+    if err != nil {
+      return err
+    }
+    m.driver.SetCPUs(c.ID, excl)
+    return nil
+  }
+
+  // Default case: assign the shared set.
+  m.driver.SetCPUs(c.ID, m.shared)
+  return nil
+}
+
+func (m *staticManager) Remove(c v1.Container, qos QoS) error {
+  m.free(m.driver.GetCPUs(c.ID))
+}
+
+func (m *staticManager) allocate(n int) (CPUList, err) {
+  excl, remaining, err := m.shared.Take(n, m.topo)
+  if err != nil {
+    return "", err
+  }
+  m.setShared(remaining)
+  return excl, nil
+}
+
+func (m *staticManager) free(c CPUList) {
+  m.setShared(m.shared.add(c))
+}
+
+func (m *staticManager) setShared(c CPUList) {
+  prev := m.shared
+  m.shared = c
+  for _, pod := range m.driver.GetPods() {
+    for _, container := range p.Containers {
+      if driver.GetCPUs(container.ID) == prev {
+        driver.SetCPUs(m.shared)
+      }
+    }
+  }
+}
+
+// @pre: container_qos = guaranteed
+func numExclusive(c v1.Container) int {
+  if c.resources.requests["cpu"] % 1000 == 0 {
+    return c.resources.requests["cpu"] / 1000
+  }
+  return 0
+}
+```
+
+##### Example pod specs and interpretation
+
+| Pod                                        | Interpretation                 |
+| ------------------------------------------ | ------------------------------ |
+| Pod [Guaranteed]:<br />&emsp;A:<br />&emsp;&emsp;cpu: 0.5 | Container **A** is assigned to the shared cpuset. |
+| Pod [Guaranteed]:<br />&emsp;A:<br />&emsp;&emsp;cpu: 2.0 | Container **A** is assigned two sibling threads on the same physical core (HT) or two physical cores on the same socket (no HT.)<br /><br /> The shared cpuset is shrunk to  make room for the exclusively allocated CPUs. |
+| Pod [Guaranteed]:<br />&emsp;A:<br />&emsp;&emsp;cpu: 1.0<br />&emsp;A:<br />&emsp;&emsp;cpu: 0.5 | Container **A** is assigned one exclusive CPU and container **B** is assigned to the shared cpuset. |
```

Comment:
nit: here and next line Pod column mentions 2 *A* containers. easy to spot on rendered page, but not here